qa_eval/evaluation.py

```diff
@@ -24,48 +26,65 @@ def evaluate_steps(
+        #actual_steps_count_total, actual_steps_error_total = defaultdict(int), defaultdict(int)
         for question in template["questions"]:
-            actual_results = chat_responses[question["id"]]
+            actual_result = responses_dict[question["id"]]
             reference_steps = question["reference_steps"]
-            if "error" in actual_results:
-                evaluation_results.append({
-                    "template_id": template_id,
-                    "question_id": actual_results["question_id"],
-                    "question_text": question["question_text"],
-                    "reference_steps": reference_steps,
+            eval_result = {
+                "template_id": template_id,
+                "question_id": actual_result["question_id"],
+                "question_text": question["question_text"],
+                "reference_steps": reference_steps,
+            }
+            if "error" in actual_result:
+                eval_result.update({
                     "status": "error",
-                    "error": actual_results["error"],
+                    "error": actual_result["error"],
                 })
+                evaluation_results.append(eval_result)
                 continue
 
-            actual_steps = actual_results["steps"]
-            score = evaluate_steps(reference_steps, actual_steps)
-
-            for step in actual_steps:
-                actual_steps_count_total[step["name"]] += 1
-                if step["status"] == "error":
-                    actual_steps_error_total[step["name"]] += 1
-
-            evaluation_results.append({
+            eval_result.update({
                 "status": "success",
-                "template_id": template_id,
-                "question_id": actual_results["question_id"],
-                "question_text": question["question_text"],
-                "reference_steps": reference_steps,
-                "actual_answer": actual_results["actual_answer"],
-                "actual_steps": actual_steps,
-                "answer_score": score,
-                "input_tokens": actual_results["input_tokens"],
-                "output_tokens": actual_results["output_tokens"],
-                "total_tokens": actual_results["total_tokens"],
-                "elapsed_sec": actual_results["elapsed_sec"],
+                "actual_steps": actual_result["steps"],
+                "actual_answer": actual_result["actual_answer"],
+                "input_tokens": actual_result["input_tokens"],
+                "output_tokens": actual_result["output_tokens"],
+                "total_tokens": actual_result["total_tokens"],
+                "elapsed_sec": actual_result["elapsed_sec"],
             })
+            if "reference_answer" in question:
+                eval_result["reference_answer"] = question["reference_answer"]
+                if not answer_evaluator:
+                    answer_evaluator = AnswerOpenAIEvaluator()
```

Comment:
Should we define a class to store this field and make this function its method?

Reply:
I think so, I don't think it's efficient to create a new object for every template and every question